### PR TITLE
[receiver/hostmetrics] Add hwscraper for hardware temperature monitoring

### PR DESCRIPTION
**Pull request by moseoh**
#### Description  
  
This PR implements a new `hwscraper` for the `hostmetricsreceiver` that collects hardware temperature data through the Linux hwmon interface. It enables monitoring of temperature sensors from various hardware components including CPU, GPU, disk, memory, and other system components.  
  
Key Features:  
  
- Hardware Temperature Collection: Collects all temperature sensor data through Linux hwmon (/sys/class/hwmon)  
- 5-Level Status System: Classifies temperature sensor states as failed, predicted_failure, degraded, needs_cleaning, or  
  ok  
- Sensor Filtering: Selective sensor collection using include/exclude patterns  
- Temperature Limits Support: Optional critical and degraded threshold metrics  
- Sanity Checking: Temperature value validation (-40°C to 200°C range)  
- Subscraper Architecture: Extensible design for future support of power, voltage, and other hardware sensors  
  
Metrics:  
  
- `hw.temperature`: Current temperature in Celsius (enabled by default)  
- `hw.temperature.limit`: Temperature thresholds (disabled by default)  
- `hw.status`: Hardware component status (ok, degraded, failed, needs_cleaning, predicted_failure)  
  
Configuration Example:  
  
```yaml  
receivers:  
  hostmetrics:  
    scrapers:  
      hw:  
        hwmon_path: "/sys/class/hwmon" # default: /sys/class/hwmon  
        temperature:  
          include:  
            sensors: [ "Composite", "Sensor 1" ] # default: .*
            match_type: strict # default: regexp
          exclude:  
            sensors: [ "Sensor 2" ]
```  
  
5-Level Temperature Status Classification:  
  
1. failed (sensor malfunction): Sensor failure or unrealistic temperature values (< -40°C or > 200°C)  
2. predicted_failure (critical): Critical threshold reached or extreme temperatures (> 85°C or < 0°C)  
3. degraded (warning): Max/Min thresholds reached or default temperatures (≥ 80°C or ≤ 5°C)  
4. needs_cleaning (attention): Maintenance required (80% of a max threshold or 75°C+)  
5. ok (normal): All conditions within normal range  
  
  
  
#### Link to tracking issue  
  
Fixes #41269  
  
  
  
#### Testing  
  
Unit Tests:  
  
- Platform-specific build tags for Linux/non-Linux test separation  
- Comprehensive test cases for temperature state determination logic  
- Sensor filtering and data parsing tests  
- Error handling and edge case testing  
  
Integration Tests:  
  
- Test data mimicking actual hwmon directory structure  
- Metrics collection and validation tests  
- Grafana dashboard visualization verification  
  
Real Environment Validation:  
  
<details>  
<summary>Prometheus Metrics Collection Results:</summary>  
  
```log  
2025-08-17T19:23:31.353+0900    debug  prometheusexporter@v0.132.0/collector.go:491   metric served: Desc{fqName: "hw_status", help: "Operational status: 1 (true) or 0 (false) for each of the possible states.", constLabels: {}, variableLabels: {id,state,type,name,parent,otel_scope_name,otel_scope_version,otel_scope_schema_url}}    {"resource": {"service.instance.id": "69f4fbb9-8df5-4a96-bfb1-cebd2f78841b", "service.name": "otelcontribcol", "service.version": "0.132.0-dev"}, "otelcol.component.id": "prometheus", "otelcol.component.kind": "exporter", "otelcol.signal": "metrics"}  
2025-08-17T19:23:31.353+0900    debug  prometheusexporter@v0.132.0/collector.go:491   metric served: Desc{fqName: "hw_temperature_celsius", help: "Temperature in degrees Celsius.", constLabels: {}, variableLabels: {id,name,parent,sensor_location,otel_scope_name,otel_scope_version,otel_scope_schema_url}} {"resource": {"service.instance.id": "69f4fbb9-8df5-4a96-bfb1-cebd2f78841b", "service.name": "otelcontribcol", "service.version": "0.132.0-dev"}, "otelcol.component.id": "prometheus", "otelcol.component.kind": "exporter", "otelcol.signal": "metrics"}  
2025-08-17T19:23:31.353+0900    debug  prometheusexporter@v0.132.0/collector.go:491   metric served: Desc{fqName: "hw_temperature_limit_celsius", help: "Temperature limit in degrees Celsius.", constLabels: {}, variableLabels: {id,limit_type,name,parent,sensor_location,otel_scope_name,otel_scope_version,otel_scope_schema_url}}  {"resource": {"service.instance.id": "69f4fbb9-8df5-4a96-bfb1-cebd2f78841b", "service.name": "otelcontribcol", "service.version": "0.132.0-dev"}, "otelcol.component.id": "prometheus", "otelcol.component.kind": "exporter", "otelcol.signal": "metrics"}  
```  
  
</details>  
  
<details>  
<summary>Grafana Dashboard Integration:</summary>  

SCR-20250804-sqxr
SCR-20250804-stfd
SCR-20250804-swei

</details>  
  
  
#### Documentation  
  
- README Update: Added hwscraper section to hostmetricsreceiver README  
- Metrics Documentation: Complete metric documentation auto-generated by mdatagen (documentation.md)  
  
Reference Standards:  
  
- https://github.com/trisch-me/semantic-conventions/blob/a46a9d88fa4c191b630d9f18fd3d6dcb5014837c/docs/hardware/temperature.md  
- https://docs.kernel.org/hwmon/sysfs-interface.html  
  
Compatibility:  
  
- Supported Platforms: Linux (kernel 2.6+ with hwmon subsystem)  
- Unsupported Platforms: macOS, Windows (graceful degradation with error return)  
- Minimum Requirements: Read access to /sys/class/hwmon directory

### CODE REVIEW

**Comment by atoulme** [comment]
Thank you for contributing this.

**Comment by moseoh** [comment]
Hi @dmitryax @braydonk, could you please take a look at this PR when you have a chance?  

It has received positive feedback from other reviewers, and since the stale bot may close it in 2 weeks, I wanted to make sure it doesn’t get lost.  

Please let me know if there are any concerns or changes needed. Thanks!

File: receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    enum: [high.critical, high.degraded, low.critical, low.degraded]
+
+  state:
+    description: The current state of the component.
+    type: string
+    enum: [degraded, failed, needs_cleaning, ok, predicted_failure]
+
+metrics:
+  hw.status:
+    enabled: true
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml]: Why is this attribute enabled given that it's based on the temperature which is already reported with `hw.temperature`? Typically we don't want to enable several metrics that provided the same information in different forms.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -249,6 +249,7 @@ receiver/hostmetricsreceiver/                                    @open-telemetry
 receiver/hostmetricsreceiver/internal/scraper/cpuscraper/        @open-telemetry/collector-contrib-approvers @dmitryax @braydonk
 receiver/hostmetricsreceiver/internal/scraper/diskscraper/       @open-telemetry/collector-contrib-approvers @dmitryax @braydonk
 receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/ @open-telemetry/collector-contrib-approvers @dmitryax @braydonk
+receiver/hostmetricsreceiver/internal/scraper/hwscraper/         @open-telemetry/collector-contrib-approvers @dmitryax @braydonk
```
Comment (dmitryax) [inline .github/CODEOWNERS]: Will you be able to help maintaining this scraper going forward?

**Review by mowies — COMMENTED** [review]
please re-generate the chloggen component list using `make generate-chloggen-components` after updating to latest main

**Comment by moseoh** [comment]
Got it, will do.

File: receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml (hunk 1 above)
Comment (moseoh) [reply receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml]: You’re right. I changed it to `enabled: false` to avoid duplicating the existing `hw.temperature` metric and reduce default cardinality.
Thanks for the feedback!

File: .github/CODEOWNERS (hunk 2 above)
Comment (moseoh) [reply .github/CODEOWNERS]: Sure, I can help with maintaining and testing the `hwscraper`.

**Comment by moseoh** [comment]
I'd like to continue working on this PR. All previous feedback has been addressed.

**Comment by paulojmdias** [comment]
@moseoh I reopened it as you would like to work on it. 👍

**Comment by moseoh** [comment]
@paulojmdias Thanks for reopening! Just to clarify — all previous review feedback has already been addressed. I've rebased onto the latest main and CI is passing. This PR is ready for review.

Also, this PR seems relevant to #44334 (SMART receiver request). The `hwscraper` introduces a subscraper architecture starting with `hw.temperature`, and the `hw.physical_disk` metrics discussed there could potentially be added as another subscraper on top of this framework.

File: .chloggen/feature-hwtemperaturescraper.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: scraper/hw
```
Comment (MovieStoreGuy) [inline .chloggen/feature-hwtemperaturescraper.yaml]: Can we please use the full name instead of an abbreviation to avoid any potential naming conflicts in future?

File: receiver/hostmetricsreceiver/internal/scraper/hardwarescraper/config.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/internal/scraper/hwscraper/internal/metadata"
+)
+
+// Config relating to HW Sensor Metric Scraper.
+type Config struct {
+	metadata.MetricsBuilderConfig `mapstructure:",squash"`
+
+	// HwmonPath specifies the path to hwmon directory
+	// Default: /sys/class/hwmon
+	HwmonPath string `mapstructure:"hwmon_path"`
```
Comment (MovieStoreGuy) [inline receiver/hostmetricsreceiver/internal/scraper/hardwarescraper/config.go:19]:
```suggestion
Path string `mapstructure:"path"`
```

This would be to avoid repeating yourself in the config considering it would currently read as:

```yaml
receivers:
  host:
    hw:
      hwmon_path: ....
```

File: receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml [hunk 5]
```diff
@@ -0,0 +1,74 @@
+type: hw
```
Comment (MovieStoreGuy) [inline receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml]: Please use the full name instead of an abbreviation

**Review by MovieStoreGuy — COMMENTED** [review]
I know you've had the code owners review this, but my only feedback would be to use `hardware` instead of `hw` so it is clear to users what it is for without needed to dive into the documentation.

File: receiver/hostmetricsreceiver/internal/scraper/hardwarescraper/config.go (hunk 4 above)
Comment (moseoh) [reply receiver/hostmetricsreceiver/internal/scraper/hardwarescraper/config.go:19]: I kept `hwmon_path` for now because the scraper reads the Linux `hwmon` interface directly, so it felt more explicit than a generic `path`. Would you still prefer `hardware.path` here?

After the rename, the config would look like this:

```yaml
receivers:
  host:
    hardware:
      hwmon_path: ...
```
I also fixed the default path handling to honor `root_path` via `gopsutilenv`, so this is now mostly an explicit override.

File: receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml (hunk 5 above)
Comment (moseoh) [reply receiver/hostmetricsreceiver/internal/scraper/hwscraper/metadata.yaml]: Updated this to use `hardware` instead of `hw`.

Since this changes the scraper type, I applied the rename consistently across the component.

**Comment by davidspek** [comment]
@moseoh is this something you’ll pick up again? What’s left to do to get this over the line?

**Comment by moseoh** [comment]
@davidspek I think the next step on my side is mostly just a rebase and chloggen regen.

This has stalled after review before and ended up closed, so I’d be happy to resume work if the maintainers/code owners are still interested in moving this feature forward.
